audioDetector.html

```diff
@@ -0,0 +1,200 @@
+var WARNING = 'WARNING' // Just used for clarity when calling displayMessage/2
+
+// Constants used to represent the state (_status) of the audio monitor
+AudioMonitor.AUDIO_DETECTED = true
+AudioMonitor.AUDIO_NOT_DETECTED = false
+
+// The minimum value of what can be considered "audio detected".
+// Can (and should) be tweaked to fit the use case and environment of the AudioMonitor.
+AudioMonitor.AUDIO_DETECTION_THRESHOLD = 0.000001
+
+/**
+ * An AudioMonitor is responsible for monitoring a DeviceSource to detect whether audio is working.
+ * It emits various events depending on the status of the DeviceSource.
+ * Only one DeviceSource can be monitored at any given time.
+ * 
+ * Usage:
+ * var audioMonitor = new AudioMonitor()
+ * audioMonitor.on('event_type', callback)
+ * 
+ * audioMonitor.monitor(someDeviceSourceWithAudio)
+ */
+function AudioMonitor () {
+  cct._EmitterMixin.call(this)
+  this._handleDeviceAccepted = this._handleDeviceAccepted.bind(this)
+  this._handleDeviceRejected = this._handleDeviceRejected.bind(this)
+  this._handleAudioMeterEvent = this._handleAudioMeterEvent.bind(this)
+  this._audioMeter = new cct.AudioMeter()
+  this._audioMeter.on('filtered', this._handleAudioMeterEvent)
+  this._status = AudioMonitor.AUDIO_DETECTED
+}
+
+AudioMonitor.prototype = {
+  constructor: AudioMonitor,
+  monitor: function (device) {
+    if (this._device) {
+      this._device.disconnect()
+    }
+    this._device = device
+    this._emit('requestingAccess')
+    device.promise.then(this._handleDeviceAccepted).catch(this._handleDeviceRejected)
+  },
+
+  // Called when DeviceSource.promise is resolved, i.e when mic access is granted.
+  _handleDeviceAccepted: function () {
+    if (webrtcDetectedType === 'plugin') {
+      // cct.AudioMeter uses the WebAudio API, which is not fully supported in Safari/Edge/IE.
+      var unsupportedError = new Error('Microphone access was accepted, but audio cannot be monitored reliably in this browser')
+      unsupportedError.name = 'UnsupportedBrowserError'
+      this._emit('error', unsupportedError)
+      this._emit(unsupportedError.name, unsupportedError)
+    } else {
+      this._device.connect(this._audioMeter)
+      this._handleAudioMeterEvent(this._audioMeter.filtered)
+    }
+  },
+
+  // Called when DeviceSource.promise is rejected, for instance if access is denied or because of some other error.
+  _handleDeviceRejected: function (error) {
+    this._emit('error', error)
+    if (error.name === 'PermissionDeniedError' || error.name === 'SecurityError' || error.name === 'NotAllowedError') {
+      // Browsers are inconsistent in what errors are thrown when mic/camera access is denied.
+      // PermissionDeniedError is thrown by Chrome and the plugin
+      // SecurityError is from Firefox 48
+      // NotAllowedError is from Firefox 50 Developer edition and onward
+      this._emit('PermissionDeniedError', error)
```

Comment:
The spec says the error should be called 'NotAllowedError', so let's use that instead.